### ISSUE TITLE
General question on if pymanopt could possibly be used for my application.

### ISSUE THREAD

**Issue by mgiglia92** [pymanopt/pymanopt#129]
Hello,

I'm going to try my best to explain the "big picture" of my application and what I have done so far, and how I can apply pymanopt to help me achieve my goals.

I have a system which has two distinct sets of dynamic equations (specifically a set of "driving" dynamics and "flying" dynamics). This system is a car/rocket from the videogame RocketLeague. For a first step (which I have already done with python GEKKO optimization library), is to minimize the time it takes for the system to drive in R2 -> transition to flying -> fly in R3 into a moving ball in minimum time. I have done this by modelling the system dynamics, and ball dynamics as a set of non-linear differential equations in time, that the solver can minimize some cost function, in this case, make the system's position and the ball's position the same in minimum time. In the future I plan on adding other methods of optimization like fuel optimal trajectories, or an optimization of hitting the ball into a certain direction, but am more focused on getting the full dynamics of the system up and running first.

The problem I'm running into is that the driving dynamics of the car are quite weird in that this car can also drive itself on the walls and ceiling of the "field". I want a way to define the "manifold" of the field (which I currently have as a list of unordered points and their associated surface normals) that way the system can drive along the manifold as an option to minimze the time it would take to get to the ball. The system would still need to be able to transition from the driving dynamics (which apply on the entire field manifold) to the flying dynamics. I'm not totally sure how to represent this in GEKKO yet as there is only a bspline function that can do f(x.y) = z, but I think I need a more general way to define and optimize along this manifold which has led me to find pymanopt.

On top of this, some of the dynamics in game are actually based on integers (on/off) and discontinuities (collisions with the ball, system collisions with the field surface, ball collisions with the field surface). Some of these I think could be approximated with continuous variables, but the less deviation from the actual system in the solution, the less I'll have to mitigate model disturbances with some sort of feedback control. Specifially, I imagine the dynamic state of the car (driving/flying) would need to be a switch of some sort, since there is not "mixing" of the dynamics, is it either totally one or the other.

So essentially I'm asking if this library has the capability of modelling and solving this complex optimization problem along with other extensions in the future. A future example I am thinking of is the following. The system has itself, the ball, and an opposing car/rocket which is the opponent system. Given the time optimal trajectories of my opponent, and the possible state spaces of the ball after my opponents collision with the ball, what trajectory should I follow to maximize my chances of intercepting the ball such that the ball's state space after both of our collisions is more in my favor than my opponents favor (maybe this could be modelled as maxmize probability of velocity of ball towards opponent goal).

Sorry for the long problem statement but I'm fairly new to the topic of topology and differential manifolds and am trying to make sure my goal is understood. 

In a sense I think if this is possible to be done here, I would like to start with reformulating the dynamics I have already with manifolds(?), such that the solvers can find solutions for what I have already setup (driving in R2, flying in R3(orientation via quaternion)). How would I go about starting that? And then where do you recommend I start with trying to represent the field itself as a manifold if I have a point cloud in R3 and associated surface normals also in R3 such that I can drive along the manifold and still feel the effects of gravity for example if I'm driving up the wall (against the gravity vector) the car should feel an acceleration from gravity.

Any guidance is appreciated, and papers are always welcome as well!

Thank you, and happy holidays!

**Comment by mgiglia92** [comment]
I think my post was a little ambitious. I want to start with something much more rudimentary.

I found the "optimal_rotations.py" example and got it running. But I'm not totally sure what I'm looking at as the output of X vs Xopt when I print it to the console.

If I wanted to find the optimal bounded control input u*(t) to drive the following simple rotational system to a final position in minimum time how should I approach writing out this problem for pymanopt?

x=[theta, theta_dot] //state of the system
u = [0; a]  // -1 < a < 1; bounded control input
x_dot = [0 1; 0 -d]x + [0; 1]u //diff eqs., 'd' is drag component

Do I imagine the state as a manifold in R2? and the input as another bounded manifold?
How does time get incorporated into this? And is there a way I should approach this so that when I move to non-linear system equations I can apply what I've learned modelling the simpler system for the more complex systems (ones maybe that include optimizing inputs that include transistioning between manifolds). I'm taking a look at An Introduction to Optimization on Smooth Manifolds by Nicolas Boumal, hopefully I'll be able to understand it haha!
